zebra-chain/src/amount.rs

```diff
@@ -262,6 +263,17 @@ pub trait Constraint {
     }
 }
 
+impl<C> ZcashDeserialize for Amount<C>
+where
+    C: Constraint,
+{
+    fn zcash_deserialize<R: std::io::Read>(
+        mut reader: R,
+    ) -> Result<Self, crate::serialization::SerializationError> {
+        Ok(reader.read_u64::<LittleEndian>()?.try_into()?)
+    }
+}
```

Comment:
This constraint seems to be causing proptest errors, maybe we need to constrain or reject invalid amounts in the proptest inputs?

Reply:
huh, thats super weird. The arbitrary impl for `Amount` should only ever generate amounts within its constraint range...

Reply:
i think its because we read as a u64, but we were serializing as an i64, I updated all the ZcashSerialization impls as well, it seems though that we have different types that are expecting the amount to be encoded as a u64 vs an i64 on the network, i don't know if this was an oversight when I wrote the amount type or if we actually represent them differently on the network contextually.

If we need to bring back the u64 based serialization i'll split this zcashdeserialize impl into two, each implemented for one of the concrete marker traits we use and hopefully the network wire format has a 1 to 1 relationship with when we use the different marker traits.

Reply:
u64 amounts are used for non-negative values, that is, almost every value in the spec:
https://github.com/ZcashFoundation/zebra/pull/870#discussion_r477642978
https://github.com/ZcashFoundation/zebra/pull/870#discussion_r477644600

i64 amounts are used for value balance changes to the Sapling transaction pool, which can be negative:

From spec 7.1 Encoding of Transactions:

Version  | Bytes  | Name  | Data Type  | Description 
-- | -- | -- | -- | --
≥4 | 8 | valueBalance | int64 | The net value of Sapling Spend transfers minus Output transfers.

It might help to have separate `Amount(u64)` and `AmountBalance(i64)` types?

Reply:
no need, we already have this in practice thanks to the parameterization, check out the updated diff, theres now two impls of ZcashSerialize and ZcashDeserialize (4 total) for Amount ^_^

Reply:
Yeah that's much nicer.